### DISCUSSION TITLE
How to initialize code-server with extensions, settings, shortcuts, etc. using config file?

### DISCUSSION THREAD

**Discussion by arita37** [coder/code-server#2842] [Q&A] [answered]
We want to initialize code-server with some-preset 
save into a yaml file on disk.

**Comment by jsjoeio** [comment]
Hi @arita37 - this is already possible, to some extent. If you look at the logs when code-server starts, it uses a config file somewhere on your machine like: `~/.config/code-server/config.yaml`

<img width="931" alt="image" src="https://user-images.githubusercontent.com/3806031/110355982-04da5800-7ff7-11eb-8a50-453e275e7108.png">

Here's what the file looks like:

```yaml
bind-addr: 127.0.0.1:8080
auth: password
password: helloworld
cert: false
```

Any flag that the CLI accepts should be configurable in the yaml file. Hope that helps! Please reopen if I misunderstood and you need something else!

**Comment by arita37** [comment]
How to pre-installed extensions ?

Suppose I have extension on disk
and wants to install them from  config file
at start

and same for dakr mode,
Shortcuts

**Comment by jsjoeio** [comment]
A few things you can do:

- you could write a script that installs extensions like so: `code-server --install-extension ms-python.python` (see [more](https://github.com/cdr/code-server/blob/main/docs/FAQ.md#where-are-extensions-stored))
- you could reuse your VS Code configuration (see [more](https://github.com/cdr/code-server/blob/main/docs/FAQ.md#how-can-i-reuse-my-vs-code-configuration)) (i.e. for dark mode or a theme) and it should work for shortcuts as well

**Comment by Ki6an** [comment]
hey,  @jsjoeio `pylint` is not working, the python version its using is  `ms-python.python v2021.2.633441544`. 
here are the versions
```python
code-server: v3.9.1
VS Code: v1.53.2
```
can you please help me fix this issue?

**Reply by jsjoeio** [reply]
Hi @Ki6an! Sorry to hear you're having issues. See #2341 - we're hoping to work on a fix soon! In the meantime, please read through that thread. There may be some workarounds.

**Reply by Ki6an** [reply]
thank you for the response, it works on the `2020.5.86806` version of python. Incase someone has the same problem I left the soln [here ](https://github.com/cdr/code-server/issues/2341#issuecomment-801450552)

**Comment by nooobcoder** [comment, marked as answer]
Use dotfiles for preconfiguring your code-server instance.
You can check my dot file config repo if you are interested. It comes with zsh by default and some pretty cool extensions.

https://github.com/nooobcoder/dot 

Please mark this discussion as answered if your query has been resolved.
Thanks

**Reply by nooobcoder** [reply]
@arita37 any update, did this work out?

**Comment by arita37** [comment]
Can I use on Colab with Colabcode ?

**Reply by nooobcoder** [reply]
Might be possible if you have a paid account to access Google colab, because Colab allows limited access to terminal on a host machine with paid accounts only. I can not assure if this dot configuration would work or not. This totally depends as I have never tried like this before. Dot files are meant to be executed on a host machine where you have greater control like a cloud compute machine that you own.

@arita37 please mark your discussion as answered if my answer was correct to your question. Colab with ColabCode is a completely different scenario with your question asked.

Thank you

**Reply by nooobcoder** [reply]
[UPDATE]

I checked out colabcode, it is nice but configuring the server with dot files is not at all possible because, you are not allowed a terminal access as I told you before. I am attaching a screenshot below. Also, it runs on a fork of CodeServer v3.7~ which is a lot older. Latest version is 3.9.2. I request you to refrain from such volatile dev environments which are pruned when the colab notebook is shut down.

It is an extremely unstable environment, it temporarily hosts a code-server using ngrok (port forwarding).

![image](https://user-images.githubusercontent.com/50350828/113508682-fd16c080-956e-11eb-9c18-ca2a1ba78343.png)

![image](https://user-images.githubusercontent.com/50350828/113508709-161f7180-956f-11eb-9e6d-171215830ec2.png)


Thank you, please mark the discussion as answered, as this does not required any further discussion.

**Reply by Ki6an** [reply]
hey, you can try [openvino-colab](https://github.com/alihussainia/openvino-colab) python library, been using it for a while & it works fine. it uses the latest version of code-server.
```python
code-server: v3.9.2
VS Code: v1.54.2
```

**Reply by nooobcoder** [reply]
@Ki6an is it persistent? After the notebook is shut down, is it preserved? If no, then it has no purpose as it becomes highly volatile to code.

**Reply by Ki6an** [reply]
use your drive as a working directory. as long as the code server is running the notebook won't disconnect (unlike it usually disconnects due to inactivity). 
one additional tip - create a `ngrok` account and use your secret `auth token` inside `server()` method get rid of frequent closing of tunnel.

**Comment by arita37** [comment]
Thanks
My question,

/install.sh requires which access right
and install ????

Am on server no sudo right

**Comment by nooobcoder** [comment]
It requires 400 to be executed

**Reply by nooobcoder** [reply]
@code-asher @jsjoeio please look into this discussion, the discussion is going off topic, even when i answered a solution to configure code-server with settings using dot files. Please mark the discussion as answered if it seems to you too that there is no discussion required to be made.

**Reply by jsjoeio** [reply]
Thanks for the tag @nooobcoder. I agree. We'll mark it as answered.
